text/1028-cli-hooks.rst

```diff
@@ -56,70 +69,184 @@ are always executed in WSL.
+  relative order.
+
+Overall the order in which the hooks are triggered is this:
+
+#. ``project.init.before``
+#. ``project.init.after`` (signifying that the project is initialized
+   and thus ready for the rest of the tools/hooks).
+#. ``branch.switch.before``
+#. ``branch.wipe.before``
+#. ``migration.apply.before``
+#. ``schema.update.before``
+#. ``branch.switch.after``
+#. ``branch.wipe.after``
+#. ``migration.apply.after``
+#. ``schema.update.after``
+
+
+Schema Hooks
+------------
+
+These hooks are a bit more abstract than the ones corresponding to specific
+CLI commands. The ``schema.update.before`` and ``schema.update.after`` hooks
+are intended for scripts that sync up the codebase with the schema state
+(regardless of what caused the schema state to change). It may be re-compiling
+the query-builder or reflected individual queries. It may be generating an ORM
+compatibility layer. It could also perform any project-specific custom
+operation.
 
 The effects produced by these hook scripts should be idempotent, so that
 running them several times on the same schema (or nearly same schema with
 trivial changes) is safe.
 
-The hooks should run *only once* per ``migration apply`` command, as long as
-some pending change was applied. They are not intended to run after applying
-every migration file if several migrations are applied by one command.
-
 The ``gel watch`` command will also trigger these hooks (without a need of a
 separate TOML config section). This is due to the fact that the ``watch``
 command conceptually monitors the schema file changes and automatically
-creates and applied migrations. So after the schema file changes are applied
-to the database, the ``migration.apply.after`` hook will be executed.
+creates and applied migrations. So the schema file changes would trigger
+``migration.apply.before``, ``schema.update.before``,
+``migration.apply.after``, and ``schema.update.after`` hooks in that relative
+order.
+
+
+Watch Configuration
+===================
+
+Sometimes in a Gel project there's a need to respond to some file changes: run
+a migration due to schema change or run code generators due to query file
+change. The CLI already supports watching the schema for changes and applying
+them, but we can generalize this mechanism and make it more flexible, much
+like the hooks that respond to CLI commands.
+
+The idea is to have a mechanism for specifying what changes you want to watch
+and a script that gets triggered by those changes. Unlike hooks, there are no
+"before" and "after" triggers here. Only one kind of response is possible:
+trigger when a change to the watched entity is detected.
+
+The watch configuration will be described in the ``gel.toml`` project manifest
+file in the ``[watch]`` table. We will introduce a ``[watch.files]`` sub-table
+for watching file system changes and triggering scripts. Potentially, we will
+also later introduce a ``[watch.gel-config]`` sub-table for monitoring changes
```

Comment:
Move ``[watch.gel-config]`` to "future extensions" or something. Probably should be just ``[watch.config]``? Or ``[watch.server-config]``?